### ISSUE TITLE
Interrupted backup not cleaning up nor resuming

### ISSUE THREAD

**Issue by bbqninja** [restic/restic#1211]
## Output of `restic version`
```
restic 0.7.1
compiled with go1.8.3 on windows/amd64
```
## How did you run restic exactly?

Notes:  

- backup directory is empty
- data directory has a 4gb iso
- Each time I interrupt I hit ctrl-c at the point ~2GB has been backed up
- The final time I let the backup complete
- This is on windows 10

```
PS C:\Users\bill\Desktop\RESTIC> $env:RESTIC_REPOSITORY = "Backup"
PS C:\Users\bill\Desktop\RESTIC> $env:RESTIC_PASSWORD = "testpass"
PS C:\Users\bill\Desktop\RESTIC> .\restic.exe init
created restic backend 0b67f0af45 at Backup

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
PS C:\Users\bill\Desktop\RESTIC> Get-ChildItem Data


    Directory: C:\Users\bill\Desktop\RESTIC\Data


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       30/05/2017   8:50 PM     4334315520 en_windows_10_multiple_editions_version_1703_updated_march_2017_x64_dvd_10189288.iso


PS C:\Users\bill\Desktop\RESTIC> Write-Host ("{0:N0}MB" -f  ((Get-ChildItem -Path Data -Recurse  | Measure-Object -property length -sum).sum / 1MB))
4,134MB
PS C:\Users\bill\Desktop\RESTIC> Write-Host ("{0:N2}MB" -f  ((Get-ChildItem -Path Backup -Recurse  | Measure-Object -property length -sum).sum / 1MB))
0.00MB
PS C:\Users\bill\Desktop\RESTIC> .\restic.exe backup Data
scan [C:\Users\bill\Desktop\RESTIC\Data]
scanned 1 directories, 1 files in 0:00
Interrupt received, cleaning up
PS C:\Users\bill\Desktop\RESTIC> Write-Host ("{0:N2}MB" -f  ((Get-ChildItem -Path Backup -Recurse  | Measure-Object -property length -sum).sum / 1MB))
2,098.37MB
PS C:\Users\bill\Desktop\RESTIC> .\restic.exe backup Data
scan [C:\Users\bill\Desktop\RESTIC\Data]
scanned 1 directories, 1 files in 0:00
Interrupt received, cleaning up
PS C:\Users\bill\Desktop\RESTIC> Write-Host ("{0:N2}MB" -f  ((Get-ChildItem -Path Backup -Recurse  | Measure-Object -property length -sum).sum / 1MB))
4,219.23MB
PS C:\Users\bill\Desktop\RESTIC> .\restic.exe backup Data
scan [C:\Users\bill\Desktop\RESTIC\Data]
scanned 1 directories, 1 files in 0:00
Interrupt received, cleaning up
PS C:\Users\bill\Desktop\RESTIC> Write-Host ("{0:N2}MB" -f  ((Get-ChildItem -Path Backup -Recurse  | Measure-Object -property length -sum).sum / 1MB))
6,379.33MB
PS C:\Users\bill\Desktop\RESTIC> .\restic.exe backup Data
scan [C:\Users\bill\Desktop\RESTIC\Data]
scanned 1 directories, 1 files in 0:00
[0:52] 100.00%  79.491 MiB/s  4.037 GiB / 4.037 GiB  2 / 2 items  0 errors  ETA 0:00
duration: 0:52, 79.44MiB/s
snapshot 36fa39fc saved
PS C:\Users\bill\Desktop\RESTIC> Write-Host ("{0:N2}MB" -f  ((Get-ChildItem -Path Backup -Recurse  | Measure-Object -property length -sum).sum / 1MB))
10,205.63MB
PS C:\Users\bill\Desktop\RESTIC>
```

## What backend/server/service did you use?

Filesystem

## Expected behavior

- backup resuming
- backup directory size no larger(ish) than sum of backed up files

## Actual behavior

- backup doesn't resume
- backup size increases by read bytes

## Steps to reproduce the behavior

See above

## Do you have any idea what may have caused this?

Cancel/resume logic not quite correct (maybe windows only?)

## Do you have an idea how to solve the issue?

Nope

**Comment by fd0** [comment]
Hey, thanks for creating the issue. It's not specific to Windows, but rather how restic (at the moment) works.

Restic reads data, splits it into chunks, bundles these chunks into so-called pack files and saves them in the repo. That are the files that you see below `data/`. After all data has been saved, an index files is written, which just contains a large list, which for each pack file lists its contents (which chunk starts at what offset). Each pack file also has a header that lists this, so the index is not strictly necessary, but this way at the start of a new backup restic reads all index files and is aware of all the data in the repo without reading all headers of all pack files.

When a backup is interrupted, restic would not have written and index, so on the next backup it looks for indexes, finds none and starts from the beginning. That's not good, and especially for remote repositories for users with small upload bandwidth that's bad.

So restic writes intermediate indexes that contain just the files that have been uploaded so far, so that when it starts the next time, it knows what has already been uploaded. The intermediate index files are written according to a few rules ([here](https://github.com/restic/restic/blob/master/internal/repository/index.go#L60-L94)) so that basically an index is written after two minutes, if it is full enough (contains at least 20 pack files) or after five minutes regardless of how many pack files have been written so far.

In your case you're using a local (=very fast) repository, so none of the conditions for an intermediate index are fulfilled, so restic does not write one. When it is run the next time, it starts from the beginning and you end up with a repo that's larger than the size of the file to backup. The next run of `restic prune` will clean this up by the way.

So, long story short: restic will resume from (roughly) where it left of, just not for such a fast backend (=local) and not for such small run times (less than a minute).

When the local cache is done (see #1040), I plan to add some kind of "journal" which keeps a local transaction log in the cache directory so that restic can pick up exactly where it left of.

I'm going to close this issue for now, please feel free to add further comments. Thanks!
